File: lib/domain/logic/category_matcher.dart
import 'dart:convert';

/// Result of applying keyword rules to a merchant string.
class CategoryGuess {
  const CategoryGuess({required this.category});
  final String category;
}

/// Keyword-based category config (bundled JSON; remote refresh in Task 22).
class CategoryConfig {
  CategoryConfig._({
    required this.version,
    required this.defaultCategory,
    required this.rules,
  });

  final String version;
  final String defaultCategory;
  final List<({String category, List<String> keywords})> rules;

  factory CategoryConfig.fromJson(Map<String, dynamic> json) {
    final rulesJson = json['rules'] as List<dynamic>? ?? [];
    final parsed = <({String category, List<String> keywords})>[];
    for (final r in rulesJson) {
      final map = r as Map<String, dynamic>;
      final category = map['category'] as String;
      final anyOf = (map['any_of'] as List<dynamic>)
          .map((e) => e as String)
          .toList();
      parsed.add((category: category, keywords: anyOf));
    }
    return CategoryConfig._(
      version: json['version'] as String? ?? '',
      defaultCategory: json['default_category'] as String? ?? 'Others',
      rules: parsed,
    );
  }

  /// Case-insensitive substring match; first rule with any keyword wins.
  CategoryGuess guessForMerchant(String merchantRaw) {
    final haystack = merchantRaw.toLowerCase();
    for (final rule in rules) {
      for (final kw in rule.keywords) {
        if (haystack.contains(kw.toLowerCase())) {
          return CategoryGuess(category: rule.category);
        }
      }
    }
    return CategoryGuess(category: defaultCategory);
  }

  /// Tries [merchantRaw] first; if that returns the default, scans the first
  /// [scanLines] lines of [ocrText] for a keyword match.
  ///
  /// Returns `(category, confidence)`:
  ///   - Merchant-name hit → 0.85
  ///   - OCR-body hit      → 0.55
  ///   - No match          → (defaultCategory, 0.10)
  ({String category, double confidence}) guessWithConfidence(
    String merchantRaw,
    String ocrText, {
    int scanLines = 15,
  }) {
    final merchantGuess = guessForMerchant(merchantRaw);
    if (merchantGuess.category != defaultCategory) {
      return (category: merchantGuess.category, confidence: 0.85);
    }

    final lines = ocrText.split(RegExp(r'\r?\n')).take(scanLines);
    for (final line in lines) {
      final haystack = line.toLowerCase();
      for (final rule in rules) {
        for (final kw in rule.keywords) {
          if (haystack.contains(kw.toLowerCase())) {
            return (category: rule.category, confidence: 0.55);
          }
        }
      }
    }

    return (category: defaultCategory, confidence: 0.10);
  }
}

File: android/app/src/main/kotlin/com/receiptdrop/receipt_drop/paymentdetect/CategoryMatcher.kt
package com.receiptdrop.receipt_drop.paymentdetect

import android.content.Context
import org.json.JSONObject

/**
 * Minimal Kotlin port of `CategoryConfig.guessWithConfidence` in
 * lib/domain/logic/category_matcher.dart (read in full before porting —
 * it's ~80 lines of pure case-insensitive substring matching, no async, no
 * LLM, nothing large enough to warrant anything but a direct port). Reads
 * the SAME bundled `assets/config/categories-v1.json` Flutter packages into
 * the APK (via `flutter_assets/`) — there is exactly one category taxonomy;
 * this is just a second, native reader of it, the same "documented port of
 * one algorithm across languages" precedent this codebase already uses for
 * parse_receipt_understanding/parseReceiptUnderstanding.
 *
 * Tiers, identical to the Dart original: 0.85 (a rule keyword found in
 * [merchantText]), 0.55 (a rule keyword found in [bodyText]'s first 15
 * lines), 0.10 (default category, no match).
 */
class CategoryMatcher private constructor(
    private val defaultCategory: String,
    private val rules: List<Pair<String, List<String>>>,
) {
    data class Guess(val category: String, val confidence: Double)

    fun guessWithConfidence(merchantText: String, bodyText: String, scanLines: Int = 15): Guess {
        val merchantHaystack = merchantText.lowercase()
        for ((category, keywords) in rules) {
            if (keywords.any { merchantHaystack.contains(it.lowercase()) }) {
                return Guess(category, 0.85)
            }
        }

        val lines = bodyText.split(Regex("""\r?\n""")).take(scanLines)
        for (line in lines) {
            val lineHaystack = line.lowercase()
            for ((category, keywords) in rules) {
                if (keywords.any { lineHaystack.contains(it.lowercase()) }) {
                    return Guess(category, 0.55)
                }
            }
        }

        return Guess(defaultCategory, 0.10)
    }

    companion object {
        private const val ASSET_PATH = "flutter_assets/assets/config/categories-v1.json"

        /** Returns null if the bundled asset can't be read/parsed — callers
         * treat that as "no category suggestion available" rather than
         * crashing (this must never block showing the overlay). */
        fun loadFromAssets(context: Context): CategoryMatcher? {
            return try {
                val json = context.assets.open(ASSET_PATH).bufferedReader(Charsets.UTF_8).use { it.readText() }
                fromJson(json)
            } catch (_: Exception) {
                null
            }
        }

        /** The actual parsing logic, split out from [loadFromAssets] so it's
         * testable in a pure JVM unit test (no Android Context needed). */
        fun fromJson(json: String): CategoryMatcher? {
            return try {
                val obj = JSONObject(json)
                val defaultCategory = obj.optString("default_category", "Others")
                val rulesJson = obj.optJSONArray("rules") ?: return CategoryMatcher(defaultCategory, emptyList())
                val rules = mutableListOf<Pair<String, List<String>>>()
                for (i in 0 until rulesJson.length()) {
                    val rule = rulesJson.getJSONObject(i)
                    val category = rule.getString("category")
                    val anyOf = rule.getJSONArray("any_of")
                    val keywords = (0 until anyOf.length()).map { anyOf.getString(it) }
                    rules.add(category to keywords)
                }
                CategoryMatcher(defaultCategory, rules)
            } catch (_: Exception) {
                null
            }
        }
    }
}
